### DISCUSSION TITLE
Performance Difference between RWX and RWO volumes

### DISCUSSION THREAD

**Discussion by I3lacx** [longhorn/longhorn#6964] [Show and tell]
Hey all,
because of some internal testing I made a couple of experiments on our Cluster related to performance of RWX and RWO volumes. Because this might be of interest to some people I thought I share it here:

The experiments were done in an artificial setting through with interesting results. I also ran some actual experiments (training a Deep Learning model), but could not observe any significant difference in this experiment. All the experiments were done in our Kubernetes Cluster consisting of 6 worker and 4 data nodes. 

The artificial experiments were done using the `fio` command and testing the **writing** speed of different operations. Different block sizes (bs) were used, as they have a drastic impact on performance. Additionally different sizes were used, mainly to keep the experiments in a doable scale.

## Overview
This below graph shows the overview of the findings. Here RWO = ReadWriteOnce, RWX = ReadWriteMany and "local" / "remote" refers to if a physical volume was present on the same physical node. E.g. when the pod is running on worker5, but the volume is located on data1, the communication still takes place, but happens through the network. `IOPS` are the I/O operations per second, whereas Bandwidth (in MiB/s) and Duration in seconds are as known. Likely the Duration is the most representative value.

**RWO vs RWX**: An important difference is that the filesystem for RWO is `ext4` and for RWX is `nfs4`. As the RWX is using a network file storage, there will be a lot of overhead, which is only necessary to allow for such a remote storage. 

![grafik](https://github.com/longhorn/longhorn/assets/27736085/fddc63c0-eab7-4de4-8649-5ce1b6ae46e8)

### Conclusion
- RWX local vs remote, does not make a difference. Likely because the same calls are being made regardless
- RWO local vs remote, for large and medium sized blocks there is a difference up to 30%, but no difference when the files get small
- RWO vs RWX: Significant Difference of 50% or more in the duration! IOPS and Bandwidth are even up to 250% faster (might be too artificial, so not soo relevant). Difference of remote and local is as expected when using the data of the left plots

# Details

When conducting artificial experiments it is important to remember:
- The transmission speed and i/o operations are heavily dependent on the block size that is read/written
- Artificial != Real
  - Experiments with artificial scenarios are **vastly** different than real runs/executions. So even a 10x difference in such a scenario might have less than 1% impact in a real scenario
  - E.g. writing to disk might be limited at a specific speed, but the actual annotation/... of the data is what actually takes time

## Setup
In order to test this, running a benchmark following: https://cloud.google.com/compute/docs/disks/benchmarking-pd-performance

- Started two notebooks with 50GB Persistent Volume, 1CPU, 2Gi Memory.
- Using different block sizes and running each experiment 10 times directly after one another
- Always 2 experiments were done at the same time

I replaced the variables with 3 values `SIZE=[6G, 2G, 400M]` and `BS=[1M, 32K, 1K]`, so 3 different experiments were performed. Each was run 10 times with a simple bash script looping over all options and preparing them in a nice format for visualization:

```
fio --name=write_throughput --directory=$TEST_DIR --numjobs=8 \
--size=$SIZE --ramp_time=2s --ioengine=libaio \
--direct=1 --verify=0 --bs=$BS --iodepth=64 --rw=write \
--group_reporting=1 --iodepth_batch_submit=64 \
--iodepth_batch_complete_max=64
```

## Detailed Results 

The results are as described in the conclusion. Here just all the results visualized as bar charts with actual values. Note the standard deviation on the bar charts. Also note that the duration should be higher for RWX, but in the comparison, the difference in duration was flipped (to fit the graph better)

![grafik](https://github.com/longhorn/longhorn/assets/27736085/42a8c781-e164-44d6-b409-d192db3e0686)
![grafik](https://github.com/longhorn/longhorn/assets/27736085/37d9311e-7979-4a19-a655-2843daa65b12)
![grafik](https://github.com/longhorn/longhorn/assets/27736085/9b4f12aa-9bf2-48a2-a8e1-9e7b4bf9fc13)
![grafik](https://github.com/longhorn/longhorn/assets/27736085/808da768-3621-4eab-84d2-76ec6d50fc40)

**Comment by derekbit** [comment]
cc @james-munson
